pkg/linter/lintRuleSyntacticRedundantCommon.go

```diff
@@ -0,0 +1,229 @@
+				// in the former case iterates only over relevant slice; in the latter skip the "myself" rule
+				var rulesToIterate []*vpcmodel.RuleOfFilter
+				if filterLayerName == vpcmodel.NaclLayer {
+					rulesToIterate = tableToRules[tableIndex][:redundantRuleIndex]
+				} else { // SG
+					rulesToIterate = tableToRules[tableIndex]
+				}
+				// gathers atomic blocks within rule's src and dst
+				srcBlocks := getAtomicBlocksOfSrcOrDst(tableAtomicBlocks, rules[redundantRuleIndex].SrcCidr)
+				dstBlocks := getAtomicBlocksOfSrcOrDst(tableAtomicBlocks, rules[redundantRuleIndex].DstCidr)
+				// iterates over cartesian product of atomic blocks within rule's src and dst; rule is redundant if all
+				// items in the cartesian product are shadowed/implies
+				ruleIsRedundant := true
+				containRules := map[int]*vpcmodel.RuleOfFilter{}
+				for _, srcAtomicBlock := range srcBlocks {
+					for _, dstAtomicBlock := range dstBlocks {
+						connOfOthers := connection.None()
+						// computes the connection of other/higher priority rules in this atomic point in the 3-dimension space
+						for otherRuleIndex, otherRule := range rulesToIterate {
+							if redundantRuleIndex == otherRuleIndex {
+								continue // relevant only to SG
+							}
+							// otherRule contributes to the shadowing/implication?
+							// namely, it contains src, dst and has a relevant connection?
+							if rules[redundantRuleIndex].IsIngress == otherRule.IsIngress &&
+								srcAtomicBlock.ContainedIn(otherRule.SrcCidr) &&
+								dstAtomicBlock.ContainedIn(otherRule.DstCidr) &&
+								!rules[redundantRuleIndex].Conn.Intersect(otherRule.Conn).IsEmpty() {
+								connOfOthers = connOfOthers.Union(otherRule.Conn)
+								if _, ok := containRules[otherRuleIndex]; !ok {
+									containRules[otherRuleIndex] = otherRule
+								}
+							}
+						}
+						// is <src, dst> shadowed/implied by other rules?
+						if !rules[redundantRuleIndex].Conn.ContainedIn(connOfOthers) {
+							ruleIsRedundant = false
+							break
+						}
+					}
+				}
+				if ruleIsRedundant {
+					res = append(res, ruleRedundant{vpcResource: config.VPC, rule: *rules[redundantRuleIndex], containRules: containRules})
+				}
+			}
+		}
+	}
+	return res, nil
+}
+
+func getAtomicBlocksOfSrcOrDst(atomicBlocks []*ipblock.IPBlock, srcOrdst *ipblock.IPBlock) []*ipblock.IPBlock {
+	res := []*ipblock.IPBlock{}
+	for _, block := range atomicBlocks {
+		if block.ContainedIn(srcOrdst) {
+			res = append(res, block)
+		}
+	}
+	return res
+}
+
+// Generates and returns two maps from tables of layer (their indexes):
+// 1. To a slice of its rules, where the location in the slice is the index of the rule
+// 2. To slice of the atomic blocks of the table
+func getTableOrientedStructs(rules []vpcmodel.RuleOfFilter) (tableToRules map[int][]*vpcmodel.RuleOfFilter,
+	tableToAtomicBlocks map[int][]*ipblock.IPBlock,
```

Comment:
suggesting the following code:
{
	tableToRules = map[int][]*vpcmodel.RuleOfFilter{}
	tableToAtomicBlocks = map[int][]*ipblock.IPBlock{}
	for i := range rules {
		filterIndex := rules[i].Filter.FilterIndex
		if _, ok := tableToRules[filterIndex]; !ok {
			tableToRules[filterIndex] = []*vpcmodel.RuleOfFilter{}
			tableToAtomicBlocks[filterIndex] = []*ipblock.IPBlock{}
		}
		tableToRules[filterIndex] = append(tableToRules[filterIndex], &rules[i])
		tableToAtomicBlocks[filterIndex] = append(tableToAtomicBlocks[filterIndex], rules[i].SrcCidr)
		tableToAtomicBlocks[filterIndex] = append(tableToAtomicBlocks[filterIndex], rules[i].DstCidr)
	}
	for tableIndex := range tableToRules {
		slices.SortFunc(tableToRules[tableIndex], func(r1, r2 *vpcmodel.RuleOfFilter) int {
			return r1.RuleIndex - r2.RuleIndex
		})
		tableToAtomicBlocks[tableIndex] = ipblock.DisjointIPBlocks(tableToAtomicBlocks[tableIndex],nil)
	}
	return tableToRules, tableToAtomicBlocks
}

Reply:
done (by you)